Fix maintenance mode

vojtechjelinek opened the pull request:
## Overview

1. This PR fixes or fixes part of #N/A.
2. This PR does the following: 

- Move maintenance mode logic from main.py to base.py.
- Refactor the maintenance page a bit.
- The`DEV_MODE` is now updated in the build.py, since we had three places from where we updated the `DEV_MODE` and all called build.py anyway.

 
## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

vojtechjelinek committed:
Fix maintenance mode

seanlip reviewed (approved):
This seems way better, thanks @vojtechjelinek!

Do we need tests for this?

seanlip commented:
@nithusha21 @ankita240796 @DubeySandeep @jameesjohn PTAL as codeowners -- this PR needs to be part of the release. Thanks!

jameesjohn reviewed (approved):
LGTM!
Thanks @vojtechjelinek

seanlip commented:
@vojtechjelinek two questions:
- Do json responses still return the html maintenance page? I remember that that was one thing I noticed. If so ... worth fixing perhaps?
- Also, let's say a user is logged in and we turn the server to maintenance mode. Will they get auto-logged out / can they log out? Or, does it not matter?

Also just noting that this PR might need backend tests to verify the routing behaviour.

Thanks!

DubeySandeep commented on core/controllers/base.py:
```diff
@@ -215,6 +215,11 @@ def dispatch(self):
                 b'https://oppiatestserver.appspot.com', permanent=True)
             return
 
+        if (feconf.ENABLE_MAINTENANCE_MODE and
```
Should we add unit test for these changes in BaseHandler test class? [Though there exist some integration test, but wanted to check whether we need to update the test as the behaviour and code-structure is a bit changed!]

vojtechjelinek replied:
Yep, I will add them.

DubeySandeep commented on main.py:
```diff
@@ -764,22 +764,5 @@ def ui_access_wrapper(self, *args, **kwargs):
-        get_redirect_route(r'%s' % feconf.ADMIN_URL, admin.AdminPage),
-        get_redirect_route(r'/adminhandler', admin.AdminHandler),
-        get_redirect_route(r'/adminrolehandler', admin.AdminRoleHandler),
-        get_redirect_route(r'/adminjoboutput', admin.AdminJobOutputHandler),
```
Are we changing the behavior here? i.e, whether this page was accessible to the admin users (or any other authorized user expert super_admins) earlier?

vojtechjelinek replied:
Hmmm, a bit, but I think that it is fine to allow only the super admins (releasers) to the page and deny everyone else.

DubeySandeep reviewed (commented):
@vojtechjelinek, Thanks for the PR, I've left a few questions, PTAL!

DubeySandeep commented:
@vojtechjelinek Note: Backend tests are failing on this PR!

seanlip reviewed (changes requested):
Forgot to add a review hold for my comments; doing that now.

ankita240796 commented on core/controllers/base.py:
```diff
@@ -215,6 +215,11 @@ def dispatch(self):
                 b'https://oppiatestserver.appspot.com', permanent=True)
             return
 
+        if (feconf.ENABLE_MAINTENANCE_MODE and
```
Break after opening bracket.

vojtechjelinek replied:
Done

ankita240796 commented on core/templates/pages/maintenance-page/maintenance-page.controller.ts:
```diff
@@ -20,17 +20,15 @@ require('domain/utilities/url-interpolation.service.ts');
     };
     ctrl.$onInit = function() {
       $scope.currentLang = 'en';
-      $rootScope.DEV_MODE = DEV_MODE;
```
Why this change?

vojtechjelinek replied:
This is no longer needed.

ankita240796 commented on main.py:
```diff
@@ -764,22 +764,5 @@ def ui_access_wrapper(self, *args, **kwargs):
 # 404 error handler (Needs to be at the end of the URLS list).
 URLS.append(get_redirect_route(r'/<:.*>', base.Error404Handler))
 
-URLS_TO_SERVE = []
-
-if (feconf.ENABLE_MAINTENANCE_MODE and
-        not current_user_services.is_current_user_super_admin()):
-    # Show only the maintenance mode page.
-    URLS_TO_SERVE = [
-        get_redirect_route(r'%s' % feconf.ADMIN_URL, admin.AdminPage),
-        get_redirect_route(r'/adminhandler', admin.AdminHandler),
-        get_redirect_route(r'/adminrolehandler', admin.AdminRoleHandler),
-        get_redirect_route(r'/adminjoboutput', admin.AdminJobOutputHandler),
-        get_redirect_route(
-            r'/admintopicscsvdownloadhandler',
-            admin.AdminTopicsCsvFileDownloader),
-        get_redirect_route(r'/<:.*>', pages.MaintenancePage)]
-else:
-    URLS_TO_SERVE = URLS
-
```
What is the purpose of this change?

vojtechjelinek replied:
The previous approach wasn't working, the `current_user_services.is_current_user_super_admin()` does not work in the main.py since we needed to move it to base.py.

ankita240796 reviewed (changes requested):
Hi @vojtechjelinek, Thanks for the PR. A few doubts, PTAL. Thanks!

vojtechjelinek committed:
Improve maintenance mode

nithusha21 reviewed (approved):
LGTM as a codeowner!

vojtechjelinek committed:
Edit build script

vojtechjelinek committed:
Fix formatting

vojtechjelinek commented:
@seanlip 

> Do json responses still return the html maintenance page? I remember that that was one thing I noticed. If so ... worth fixing perhaps?

Fixed this, by using the exception handler for the maintenance mode. The benefit is that the responses now also return 503 so that it is more HTTP appropriate.

> Also, let's say a user is logged in and we turn the server to maintenance mode. Will they get auto-logged out / can they log out? Or, does it not matter?

Nope they won't, but I have modified the code so that there is no interaction with the user models at all.

vojtechjelinek committed:
Fix code

seanlip commented on core/controllers/base.py:
```diff
@@ -119,6 +119,9 @@ class PageNotFoundException(Exception):
         """Error class for an internal server side error (error code 500)."""
 
+    class TemporaryMaintenanceException(Exception):
+        """Error class for a temporary maintenance (error code 503)."""
```
Let's explain this a bit more:

Error class for when the server is currently down for temporary maintenance (error code 503).

vojtechjelinek replied:
Done

seanlip commented on core/controllers/base.py:
```diff
@@ -215,6 +220,15 @@ def dispatch(self):
                 b'https://oppiatestserver.appspot.com', permanent=True)
             return
 
+        if feconf.ENABLE_MAINTENANCE_MODE and not self.is_super_admin:
```
Just a question about this. Suppose the user is a super admin but is not logged in (or gets logged out somehow), and the server is currently in maintenance mode. They'll need to log in somehow. Will they be able to do so?

(Perhaps sufficient for them to just go to /admin, but I just wanted to check that that would still work even if the user isn't logged in already. Probably should have a backend integration test for this case. And check that, if a non-super-admin tries to log in there, they still can't get access to anything.)

vojtechjelinek replied:
They can't go to admin but can use a different URL to login.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -211,6 +211,23 @@ def test_requests_for_invalid_paths(self):
+        with swap_maintenace_mode:
+            response = (
+                self.get_html_response('/library', expected_status_int=503))
+            self.assertIn('htmlWebpackPlugin.options.statusCode', response.body)
```
Should you be checking for the existence of "503" too, somewhere? This assertion doesn't really seem to be testing anything related to the 503 itself.

vojtechjelinek replied:
On line 219 there is `expected_status_int=503`.

seanlip replied:
Also perhaps add an assertNotIn that corresponds to line 229 below.

vojtechjelinek replied:
Done.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -211,6 +211,23 @@ def test_requests_for_invalid_paths(self):
+            feconf, 'ENABLE_MAINTENANCE_MODE', True)
+        login_super_admin = self.login_context(
+            self.SUPER_ADMIN_EMAIL, is_super_admin=True)
+        with swap_maintenace_mode, login_super_admin:
```
maintenance is misspelled.

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -179,6 +189,27 @@ def generate_app_yaml(deploy_mode=False):
+        common.CONSTANTS_PATH, r'"DEV_MODE": .*', dev_mode_variable)
+
+    enable_maintenance_mode_variable = (
+        'ENABLE_MAINTENANCE_MODE = ' + python_utils.UNICODE(maintenance_mode))
```
Please use '%s' interpolation rather than concatenation, for consistency (here and elsewhere).

vojtechjelinek replied:
Done.

seanlip commented on scripts/release_scripts/deploy.py:
```diff
@@ -68,6 +68,8 @@
 _PARSER.add_argument(
     '--version', help='version to deploy', type=str)
+_PARSER.add_argument(
+    '--maintenance_mode', action='store_true', default=False)
```
Update instructions at top of file for how to use this.

vojtechjelinek replied:
Done.

seanlip commented on scripts/run_e2e_tests.py:
```diff
@@ -271,7 +256,6 @@ def build_js_files(dev_mode_setting, deparallelize_terser=False):
         deparallelize_terser: bool. Represents whether to use webpack
         compilation config that disables parallelism on terser plugin.
     """
-    update_dev_mode_in_constants_js(CONSTANT_FILE_PATH, dev_mode_setting)
```
Shouldn't anything else be changed in this file? (e.g. if we want to run the e2e tests in dev vs prod)?

Or are we forcing one mode from here on? /cc @U8NWXD

vojtechjelinek replied:
Nope, the `DEV_MODE` is now updated in the build.py, I decided to refactor this for a bit since we had three places from where we updated the `DEV_MODE` and all called build.py anyway. I'm adding this to the PR description.

seanlip reviewed (changes requested):
Thanks @vojtechjelinek! Took an initial pass.

ankita240796 reviewed (approved):
LGTM, Thanks @vojtechjelinek!

seanlip commented on core/controllers/base_test.py:
```diff
@@ -211,6 +211,23 @@ def test_requests_for_invalid_paths(self):
         self.delete_json('/library/data', expected_status_int=404)
 
+    def test_show_maintenance_mode_when_enabled(self):
+        swap_maintenace_mode = self.swap(
```
spelling: maintenance. I'm seeing this in multiple places -- please fix.

vojtechjelinek replied:
Done.

seanlip commented on core/controllers/base_test.py:
```diff
@@ -211,6 +211,23 @@ def test_requests_for_invalid_paths(self):
 
         self.delete_json('/library/data', expected_status_int=404)
 
+    def test_show_maintenance_mode_when_enabled(self):
```
Can you please add comprehensive tests for both json and html handlers?

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -149,24 +149,34 @@
+        deploy_mode: bool. Whether the script is being called from deploy
+            script.
+        maintenance_mode: bool. Whether the site should be put into
+            the maintenance mode.
```
Drop "the"

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -149,24 +149,34 @@
 _PARSER.add_argument(
```
A bit unrelated to your PR, but what does this arg mean? What is terser? Maybe add a description/docstring?

vojtechjelinek replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1257,25 +1288,38 @@ def generate_build_directory(hashes):
-                build_using_webpack(WEBPACK_PROD_CONFIG)
-            generate_app_yaml(deploy_mode=options.deploy_mode)
-            generate_build_directory(hashes)
+        hashes = generate_hashes()
```
The removal of "if not options.minify_third_party_libs_only" from here seems to change the behaviour, since now hashes are always computed. Why do we do this? Or is the point that the line before this should be removed? I'm somewhat confused...

vojtechjelinek replied:
This is now handled on lines 1302. The issue is that in the frontend tests we don't set the `DEV_MODE` even in prod mode and just minify the third party libs.

seanlip commented on scripts/release_scripts/deploy.py:
```diff
@@ -147,21 +147,15 @@ def preprocess_release(app_name, deploy_data_path):
-    # Changes the DEV_MODE constant in assets/constants.ts.
     with python_utils.open_file(
-        os.path.join(CONSTANTS_FILE_PATH), 'r') as assets_file:
+        os.path.join(common.CONSTANTS_PATH), 'r') as assets_file:
```
CONSTANTS_FILE_PATH is a better name

vojtechjelinek replied:
Done.

seanlip commented on scripts/release_scripts/deploy.py:
```diff
@@ -215,17 +209,22 @@ def update_and_check_indexes(app_name):
+    """Builds and minifies all the scripts.
+
+    Args:
+        maintenance_mode: bool. Whether enable the maintenance mode.
```
Whether --> Whether to

vojtechjelinek replied:
Done.

vojtechjelinek committed:
Address comments

vojtechjelinek committed:
Fix tests

vojtechjelinek committed:
Remove login params from app_dev.yaml

vojtechjelinek committed:
Merge branch 'develop' of github.com:oppia/oppia into fix-maintenance-mode

kevintab95 commented:
Hi @vojtechjelinek is this ready for another round of review? If so, please assign the reviewers and @-mention for review. Thanks!

vojtechjelinek commented:
@seanlip @DubeySandeep PTAL

seanlip commented:
@vojtechjelinek please make sure you reply to all the comments.

seanlip reviewed (approved):
LGTM. Thanks @vojtechjelinek!

DubeySandeep commented on core/templates/pages/maintenance-page/maintenance-page.controller.ts:
```diff
@@ -20,17 +20,15 @@ require('domain/utilities/url-interpolation.service.ts');
+  '$scope', 'DocumentAttributeCustomizationService', 'UrlInterpolationService',
+  function(
+      $scope, DocumentAttributeCustomizationService, UrlInterpolationService) {
+    let ctrl = this;
```
Any specific reason for using both $scope and `ctrl` here?

DubeySandeep replied:
Ignore this comment*

DubeySandeep commented on scripts/build.py:
```diff
@@ -147,26 +147,46 @@
-                prod_file_prefix + file_path)
+            if maintenance_mode:
+                content = content.replace(
+                    file_path, prod_file_prefix + maintenance_page_path)
```
Are we going to serve assets (images/videos), js and ts files in maintenance mode?

URLs like: https://www.oppia.org/assets/images/avatar/oppia_avatar_100px.svg https://www.oppia.org/templates/css/oppia.css https://www.oppia.org/third_party/generated/js/third_party.min.js

seanlip replied:
For quick review, just wanted to confirm this is fine either way. The main thing to ensure in maintenance mode is that no *changes* can be made to the system.

DubeySandeep reviewed (commented):
@vojtechjelinek, The changes LGTM for the code owner files, I have one question, PTAL!

DubeySandeep reviewed (approved):
Thanks for the quick response @seanlip! 

LGTM!
